File: Lab4/app/src/main/java/com/example/lab4/MainActivity.java
package com.example.lab4;

import android.app.Activity;
import android.icu.text.DecimalFormat;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

public class MainActivity extends Activity {
    EditText edtdoC, edtdoF;
    Button btncf, btnfc, btnclear;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        edtdoC = findViewById(R.id.edtC);
        edtdoF = findViewById(R.id.edtF);
        btncf = findViewById(R.id.btnCF);
        btnfc = findViewById(R.id.btnFC);
        btnclear = findViewById(R.id.btnclear);
        btncf.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DecimalFormat dcf = new DecimalFormat("#.00");
                String doC = edtdoC.getText() + "";
                int C = Integer.parseInt(doC);
                edtdoF.setText("" + dcf.format(C * 1.8 + 32));
            }
        });
        btnfc.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DecimalFormat dcf = new DecimalFormat("#.00");
                String doF = edtdoF.getText().toString();
                int F = Integer.parseInt(doF);
                double C = (F - 32) / 1.8;
                edtdoC.setText(dcf.format(C));
            }
        });
        btnclear.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                edtdoC.setText("");
                edtdoF.setText("");
            }
        });
    }
}

File: Lab16/app/src/main/java/com/example/lab16/MainActivity.kt
package com.example.lab16

import android.os.Bundle
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import android.content.SharedPreferences
import android.widget.Button
import android.widget.EditText
import android.widget.TextView

class MainActivity : AppCompatActivity() {
    private lateinit var edta: EditText
    private lateinit var edtb: EditText
    private lateinit var edtkq: EditText
    private lateinit var btntong: Button
    private lateinit var btnclear: Button
    private lateinit var txtlichsu: TextView

    private var lichsu: String = ""
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }
        edta = findViewById(R.id.edta)
        edtb = findViewById(R.id.edtb)
        edtkq = findViewById(R.id.edtkq)
        btntong = findViewById(R.id.btntong)
        btnclear = findViewById(R.id.btnclear)
        txtlichsu = findViewById(R.id.txtlichsu)

        // Lấy lại dữ liệu trong SharedPreferences
        val myprefs = getSharedPreferences("mysave", MODE_PRIVATE)
        lichsu = myprefs.getString("ls", "") ?: ""
        txtlichsu.text = lichsu

        btntong.setOnClickListener {
            val a = edta.text.toString().toIntOrNull() ?: 0
            val b = edtb.text.toString().toIntOrNull() ?: 0
            val kq = a + b
            edtkq.setText(kq.toString())
            lichsu += "$a + $b = $kq\n"   // thêm lịch sử + xuống dòng
            txtlichsu.text = lichsu
        }

        btnclear.setOnClickListener {
            lichsu = ""
            txtlichsu.text = lichsu
        }
    }
    override fun onPause() {
        super.onPause()
        val myprefs = getSharedPreferences("mysave", MODE_PRIVATE)
        myprefs.edit()
            .putString("ls", lichsu)
            .apply() // dùng apply() cho bất đồng bộ, thay cho commit()
    }
}